subsys/bluetooth/audio/aics.c

```diff
@@ -164,8 +167,46 @@ static ssize_t read_input_status(struct bt_conn *conn,
 				 sizeof(inst->srv.status));
 }
 
+static void state_notify_work_handler(struct k_work *work)
+{
+	struct k_work_delayable *d_work = k_work_delayable_from_work(work);
+	struct bt_aics *inst = CONTAINER_OF(d_work, struct bt_aics, srv.state_notify_work);
+	int err;
+
+	err = bt_gatt_notify_uuid(NULL, BT_UUID_AICS_STATE, inst->srv.service_p->attrs,
+				  &inst->srv.state, sizeof(inst->srv.state));
+	if (err < 0 && err != -ENOTCONN) {
```

Comment:
`bt_gatt_notify_uuid` seems to be a blocking operation w.r.t. getting buffers when sending notifications (see `bt_att_chan_create_pdu`), and will only return an error if something besides buffers failed, in which case, retrying won't help. 

Does it make sense to retry at all? Or is this based on the assumption that the buffer handling will change?

Reply:
With the addition of meta_data to ATT buffers, the meta_data is free'd from sysworkq, thus waiting in sysworkq is not allowed because it may cause a deadlock. I've added a log locally that shows the thread from which the meta_data is free'd:
```

[00:00:28.687,200] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:28.785,500] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:28.932,900] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.031,200] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.129,500] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.179,000] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.227,800] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.277,300] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.326,200] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.327,100] <err> bt_att: tx_meta_data_free called from BT RX
[00:00:29.375,600] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.424,400] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.473,900] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.522,700] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.572,300] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.621,000] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.670,600] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.719,300] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.768,900] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.817,600] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.867,200] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.915,900] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.965,100] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:29.965,300] <err> bt_att: tx_meta_data_free called from BT RX
[00:00:30.014,200] <err> bt_att: tx_meta_data_free called from sysworkq
[00:00:30.063,400] <err> bt_att: tx_meta_data_free called from sysworkq
```

In other words, the `bt_gatt_notify` might return `-ENOMEM` when called from BT RX and sysworkq

Reply:
I see. Thanks for clarifying and investigating. Do you know if this is intentional, or something that should be fixed in GATT? 

I do wonder though, if we should only retry on `-ENOMEM` then? Are there other error codes that can be returned that makes sense to retry on?

Reply:
It's intentional. `tx_complete_work` in conn.c is executed from sysworkq.
Maybe we could have a dedicated le audio thread in which we could wait/sleep/etc bot blocking the rest of the system. Unfortunately, the GATT API uses flat buffers (instead of net_buf's that could be referenced and put to fifo and processed from other thread)

Reply:
> It's intentional. tx_complete_work in conn.c is executed from sysworkq.

Do you know if it is documented anywhere? We should be wary about assuming undocumented behavior, but in any case, this PR seems to handle it fine regardless

Reply:
It is not. I had to debug.
It was introduced some time ago (in 2019)
Bluetooth: conn: Defer bt_conn_tx callback to system wq
8a7615f4a3d8932481277dffd41a4a72febd0a61